ecosystem/ffmpeg_plugin/kahawai_enc.c

```diff
@@ -0,0 +1,396 @@
+    s->dev_handle = shared_st_handle;
+    av_log(ctx, AV_LOG_VERBOSE, "use shared st_handle 0x%" PRIx64 "\n",
+           (unsigned long)shared_st_handle);
+  }
+  ++active_session_cnt;
+
+  ops_tx.name = "st20p";
+  ops_tx.priv = s;                // Handle of priv_data registered to lib
+  ops_tx.port.payload_type = 112; // TX_ST20_PAYLOAD_TYPE
+  ops_tx.device = ST_PLUGIN_DEVICE_AUTO;
+  ops_tx.notify_frame_available = tx_st20p_frame_available;
+  ops_tx.framebuff_cnt = s->fb_cnt;
+  ops_tx.notify_frame_done = tx_st20p_frame_done;
+  //   ops_tx.flags |= ST20_TX_FLAG_USER_TIMESTAMP;
+
+  pthread_mutex_init(&(s->get_frame_mutex), NULL);
+  pthread_cond_init(&(s->get_frame_cond), NULL);
+
+  av_log(ctx, AV_LOG_VERBOSE, "st20p_tx_create st_handle 0x%" PRIx64 "\n",
+         (unsigned long)s->dev_handle);
+  av_log(ctx, AV_LOG_VERBOSE, "udp_port %d\n", s->udp_port);
+
+  s->tx_handle = st20p_tx_create(s->dev_handle, &ops_tx);
+  if (!s->tx_handle) {
+    av_log(ctx, AV_LOG_ERROR, "st20p_tx_create failed\n");
+    return AVERROR(EIO);
+  }
+
+  s->output_frame_size = st20p_tx_frame_size(s->tx_handle);
+
+  if (s->output_frame_size <= 0) {
+    av_log(ctx, AV_LOG_ERROR, "st20p_tx_frame_size failed\n");
+    return AVERROR(EINVAL);
+  }
+
+  av_log(ctx, AV_LOG_VERBOSE, "st20p_tx_create finished\n");
+
+  s->frame_counter = 0;
+  s->frame = NULL;
+
+  frame_tx_completed = 0;
+  tx_completed = false;
+  return 0;
+}
+
+static int kahawai_write_packet(AVFormatContext *ctx, AVPacket *pkt) {
+  KahawaiMuxerContext *s = ctx->priv_data;
+  int ret = 0;
+
+  int i = 0, j = 0, h = 0;
+  av_log(ctx, AV_LOG_VERBOSE, "kahawai_write_packet triggered\n");
+  s->frame = st20p_tx_get_frame(s->tx_handle);
+  if (!s->frame) {
+    pthread_mutex_lock(&(s->get_frame_mutex));
+    pthread_cond_wait(&(s->get_frame_cond), &(s->get_frame_mutex));
+    pthread_mutex_unlock(&(s->get_frame_mutex));
+
+    s->frame = st20p_tx_get_frame(s->tx_handle);
+    if (!s->frame) {
+      av_log(ctx, AV_LOG_ERROR, "st20p_tx_get_frame failed\n");
+      return AVERROR(EIO);
+    }
+  }
+  av_log(ctx, AV_LOG_VERBOSE, "st20p_tx_get_frame: 0x%" PRIx64 "\n",
+         (unsigned long)(s->frame->addr[0]));
+
+  if (s->frame->data_size != s->output_frame_size) {
+    av_log(ctx, AV_LOG_ERROR,
+           "Unexpected frame size received: %" PRIu64 " (%" PRIu64
+           " expected)\n",
+           s->frame->data_size, s->output_frame_size);
+    return AVERROR(EIO);
+  }
+
+  uint8_t *data[4];
+  int linesize[4];
+  av_image_fill_arrays(data, linesize, pkt->data, s->pixel_format, s->width,
+                       s->height, 1);
+  av_image_copy(s->frame->addr, s->frame->linesize, data, linesize,
+                s->pixel_format, s->width, s->height);
+
+  st20p_tx_put_frame(s->tx_handle, s->frame);
+  av_log(ctx, AV_LOG_VERBOSE, "st20p_tx_put_frame: 0x%" PRIx64 "\n",
+         (unsigned long)(s->frame->addr[0]));
+
+  return 0;
+}
+
+static int kahawai_write_trailer(AVFormatContext *ctx) {
```

Comment:
I don't find anyone called this function, is it a dead code?

Reply:
Fixed. Added kahawai_write_trailer as write_trailer callback to kahawai_muxer